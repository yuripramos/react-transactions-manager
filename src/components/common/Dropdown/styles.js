import styled, { css } from "styled-components";
import { media, rem } from "../../../styles/tools";
import {
  white,
  grey30,
  black50,
  darkBlue,
  green,
  red
} from "../../../styles/settings";
import { Label } from "../../common/Input/styles";

export const DropdownField = styled.select`
  width: 100%;
  height: ${rem(48)};
  border-radius: ${rem(4)};
  background-color: ${white};
  border: solid ${rem(1)} ${grey30};
  font-family: Lato;
  font-size: ${rem(13)};
  letter-spacing: ${rem(0.8)};
  color: ${black50};
  padding: ${rem(15)} ${rem(55)} 0 ${rem(6.4)};
  transition: 0.3s padding, 0.3s border;
  box-shadow: none;

  ::-ms-clear {
    display: none !important;
  }

  :focus {
    outline: none;
    border: solid ${rem(1)} ${darkBlue};
  }

  ${({ isEmpty, tinyLabels }) => !isEmpty && tinyLabels && revealContent2};
  :not(:focus) {
    ${({ tinyLabels, isEmpty }) => tinyLabels && !isEmpty && revealContent3};
  }

  ${({ hasLabel }) =>
    !hasLabel &&
    css`
      height: auto;
      padding: ${rem(8)} 0 ${rem(8)} ${rem(15)};
      border-radius: ${rem(20)};
      appearance: none;
    `};
`;

export const DropdownWrapper = styled.div`
  letter-spacing: ${rem(0.1)};
  display: inline-block;
  position: relative;
  margin: 0 ${rem(6)} ${rem(6)} 0;

  select:not(:focus) {
    ${({ valid }) =>
      valid === false &&
      css`
        border-color: ${red};
      `};
  }

  ${({ width }) =>
    width &&
    css`
      width: ${width}%;
    `};
`;

const revealContent2 = css`
  ~ ${Label} {
    top: ${rem(10)};
    font-family: Lato Bold;
    font-size: ${rem(11)};
    letter-spacing: ${rem(0.1)};
    text-align: left;
    color: ${darkBlue};
  }
`;

const revealContent3 = css`
  ~ ${Label} {
    top: ${rem(12)};
    color: ${black50};
  }
`;

export const IconContainer = styled.div`
  position: absolute;
  top: 50%;
  left: 0;
  width: ${rem(55)};
  display: flex;
  justify-content: center;
  align-items: center;
  transform: translateY(-50%);
  color: ${black50};
  transition: 0.3s color;
  pointer-events: none;
`;
