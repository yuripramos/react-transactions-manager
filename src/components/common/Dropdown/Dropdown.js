import React, { Component } from "react";
import { string, oneOfType, func, bool, number } from "prop-types";
import { DropdownWrapper, DropdownField } from "./styles";
import { Label } from "../../common/Input/styles";

class Dropdown extends Component {
  constructor(props) {
    super(props);

    this.state = {
      emptyState: false,
      isValid: typeof props.valid === "boolean" ? props.valid : null
    };
  }

  componentDidUpdate(prevProps) {
    if (this.props.value === "" && this.state.isValid !== null) {
      this.setState({
        isValid: null
      });
    }

    if (prevProps.value === "" && this.props.value !== "") {
      this.setState({
        isValid: this.props.valid ? this.props.valid() : this.state.isValid
      });
    }
  }

  render() {
    const {
      value,
      placeholder,
      onChange,
      type,
      name,
      dataTest,
      list,
      label,
      tinyLabels,
      width,
      valid,
      noEmptySelection
    } = this.props;

    const { emptyState, isValid } = this.state;
    const emptyOptionTemplate = noEmptySelection
      ? []
      : [{ name: "", selected: true }];

    return (
      <DropdownWrapper valid={emptyState ? false : isValid} width={width}>
        <DropdownField
          required
          onChange={e => {
            onChange && onChange(e);
            this.setState({ emptyState: e.target.value === "" });
          }}
          onBlur={() => {
            this.setState({
              isValid: valid ? valid() : this.state.isValid
            });
          }}
          type={type}
          value={value}
          name={name}
          placeholder={placeholder}
          data-test={dataTest ? dataTest : Dropdown.displayName}
          isEmpty={!value}
          tinyLabels={tinyLabels}
        >
          {emptyOptionTemplate.concat(list).map((e, i) => (
            <option key={i} value={e.value} defaultValue={e.selected}>
              {e.name}
            </option>
          ))}
        </DropdownField>
        {label && <Label tinyLabels={tinyLabels}>{label}</Label>}
      </DropdownWrapper>
    );
  }
}

Dropdown.defaultProps = {
  value: "",
  name: "",
  label: ""
};

Dropdown.propTypes = {
  placeholder: string,
  type: string,
  name: string,
  label: string,
  tinyLabels: bool,
  value: oneOfType([number, string]),
  valid: oneOfType([bool, func]),
  onChange: func
};

export default Dropdown;
