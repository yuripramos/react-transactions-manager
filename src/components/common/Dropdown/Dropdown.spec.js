import React from "react";
import { shallow } from "enzyme";

import Dropdown from "./Dropdown";

import { spy } from "sinon";

const props = {
  onChange: () => {},
  name: "dropdown",
  placeholder: "Dropdown",
  valid: () => true,
  list: [{ name: "one" }, { name: "two" }],
  value: "",
  label: "label"
};

describe("Dropdown component", () => {
  it("should match snapshot with no validation and without accountNumber", () => {
    const wrapper = shallow(<Dropdown accountNumber={false} {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should match snapshot with no validation and with accountNumber", () => {
    const wrapper = shallow(<Dropdown accountNumber={true} {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should match snapshot with prop valid as boolean", () => {
    const props = {
      onChange: () => {},
      name: "dropdown",
      placeholder: "Dropdown",
      valid: false,
      list: [{ name: "one" }, { name: "two" }],
      value: ""
    };

    const wrapper = shallow(<Dropdown accountNumber={true} {...props} />);
    expect(wrapper).toMatchSnapshot();
  });


  it("should match snapshot with no prop valid", () => {
    const props = {
      onChange: () => {},
      name: "dropdown",
      placeholder: "Dropdown",
      list: [{ name: "one" }, { name: "two" }],
      value: ""
    };

    const wrapper = shallow(<Dropdown accountNumber={true} {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should properly call componentDidUpdate", () => {
    spy(Dropdown.prototype, "componentDidUpdate");

    const wrapper = shallow(<Dropdown accountNumber={true} {...props} />);
    wrapper.setProps({ value: "two" });

    expect(Dropdown.prototype.componentDidUpdate.calledOnce);
    expect(wrapper.state()).toEqual({
      emptyState: false,
      isValid: true
    });

    Dropdown.prototype.componentDidUpdate.restore();
  });

  it("should properly call componentDidUpdate with no valid prop", () => {
    const props = {
      onChange: () => {},
      name: "dropdown",
      placeholder: "Dropdown",
      list: [{ name: "one" }, { name: "two" }],
      value: "",
      label: "label"
    };

    spy(Dropdown.prototype, "componentDidUpdate");

    const wrapper = shallow(<Dropdown accountNumber={true} {...props} />);
    wrapper.setProps({ value: "two" });

    expect(Dropdown.prototype.componentDidUpdate.calledOnce);
    expect(wrapper.state()).toEqual({
      emptyState: false,
      isValid: null
    });

    Dropdown.prototype.componentDidUpdate.restore();
  });
});
