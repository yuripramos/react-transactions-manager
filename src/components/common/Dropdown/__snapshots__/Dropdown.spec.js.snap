// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown component should match snapshot with data-test attribute 1`] = `
<styled.div
  valid={false}
>
  <styled.select
    data-test="dropdown"
    hasLabel=""
    isEmpty={true}
    name="dropdown"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="Dropdown"
    required={true}
    value=""
  >
    <option
      defaultValue={true}
      key="0"
    />
    <option
      defaultValue={false}
      key="1"
    />
    <option
      defaultValue={false}
      key="2"
    />
  </styled.select>
  <styled.div>
    <Icon
      color={null}
      cursor={false}
      fixCrossAxisAlign={null}
      height="24"
      type="Attention"
      width="24"
    />
  </styled.div>
</styled.div>
`;

exports[`Dropdown component should match snapshot with emptyState 1`] = `
<styled.div
  valid={false}
>
  <styled.select
    hasLabel="label"
    isEmpty={true}
    name="dropdown"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="Dropdown"
    required={true}
    value=""
  >
    <option
      defaultValue={true}
      key="0"
    />
    <option
      defaultValue={false}
      key="1"
    />
    <option
      defaultValue={false}
      key="2"
    />
  </styled.select>
  <styled.label>
    label
  </styled.label>
  <styled.div>
    <Icon
      color={null}
      cursor={false}
      fixCrossAxisAlign={null}
      height="24"
      type="Attention"
      width="24"
    />
  </styled.div>
</styled.div>
`;

exports[`Dropdown component should match snapshot with no prop valid 1`] = `
<styled.div
  valid={null}
>
  <styled.select
    hasLabel=""
    isEmpty={true}
    name="dropdown"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="Dropdown"
    required={true}
    value=""
  >
    <option
      defaultValue={true}
      key="0"
    />
    <option
      defaultValue={false}
      key="1"
    />
    <option
      defaultValue={false}
      key="2"
    />
  </styled.select>
</styled.div>
`;

exports[`Dropdown component should match snapshot with no validation and with accountNumber 1`] = `
<styled.div
  valid={null}
>
  <styled.select
    hasLabel="label"
    isEmpty={true}
    name="dropdown"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="Dropdown"
    required={true}
    value=""
  >
    <option
      defaultValue={true}
      key="0"
    />
    <option
      defaultValue={false}
      key="1"
    />
    <option
      defaultValue={false}
      key="2"
    />
  </styled.select>
  <styled.label>
    label
  </styled.label>
</styled.div>
`;

exports[`Dropdown component should match snapshot with no validation and without accountNumber 1`] = `
<styled.div
  valid={null}
>
  <styled.select
    hasLabel="label"
    isEmpty={true}
    name="dropdown"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="Dropdown"
    required={true}
    value=""
  >
    <option
      defaultValue={true}
      key="0"
    />
    <option
      key="1"
    >
      one
    </option>
    <option
      key="2"
    >
      two
    </option>
  </styled.select>
  <styled.label>
    label
  </styled.label>
</styled.div>
`;

exports[`Dropdown component should match snapshot with prop valid as boolean 1`] = `
<styled.div
  valid={false}
>
  <styled.select
    hasLabel=""
    isEmpty={true}
    name="dropdown"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="Dropdown"
    required={true}
    value=""
  >
    <option
      defaultValue={true}
      key="0"
    />
    <option
      defaultValue={false}
      key="1"
    />
    <option
      defaultValue={false}
      key="2"
    />
  </styled.select>
  <styled.div>
    <Icon
      color={null}
      cursor={false}
      fixCrossAxisAlign={null}
      height="24"
      type="Attention"
      width="24"
    />
  </styled.div>
</styled.div>
`;

exports[`Dropdown component should match snapshot with prop valid as func 1`] = `
<styled.div
  valid={null}
>
  <styled.select
    hasLabel="label"
    isEmpty={true}
    name="dropdown"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="Dropdown"
    required={true}
    value=""
  >
    <option
      defaultValue={true}
      key="0"
    />
    <option
      defaultValue={false}
      key="1"
    />
    <option
      defaultValue={false}
      key="2"
    />
  </styled.select>
  <styled.label>
    label
  </styled.label>
</styled.div>
`;
